app/models/course_option.rb

```diff
@@ -12,12 +12,27 @@ class CourseOption < ApplicationRecord
   }
 
   enum vacancy_status: {
-    both_full_time_and_part_time_vacancies: 'both_full_time_and_part_time_vacancies',
-    part_time_vacancies: 'part_time_vacancies',
-    full_time_vacancies: 'full_time_vacancies',
+    vacancies: 'vacancies',
     no_vacancies: 'no_vacancies',
   }
 
+  def update_vacancy_status_from_detailed_description!(description)
+    no_vacancies! and return if description == 'no_vacancies'
+    vacancies! and return if description == 'both_full_time_and_part_time_vacancies'
```

Comment:
I had a little trouble following the logic here initially, I think because of the mixing of guard clauses and full `if` statements. Is there a way to write it differently, perhaps a case statement? That would allow you to remove the `and return`.

Also I've never seen the `vacancies!` API used in an enum, not sure if it's worth the suprise vs `update!(vacancy_status: 'vacancies')`?

Reply:
I like the case statement idea. WDYT about this:

```
    case description
    when 'no_vacancies'
      no_vacancies!
    when 'both_full_time_and_part_time_vacancies'
      vacancies!
    when 'full_time_vacancies'
      full_time? ? vacancies! : no_vacancies!
    when 'part_time_vacancies'
      part_time? ? vacancies! : no_vacancies!
    else
      raise InvalidVacancyStatusDescriptionError, description
    end
```

And a variant without the enum methods:

```
    case description
    when 'no_vacancies'
      no_vacancies!
    when 'both_full_time_and_part_time_vacancies'
      vacancies!
    when 'full_time_vacancies'
      full_time? ? update(vacancy_status: :vacancies) : update(vacancy_status: :no_vacancies)
    when 'part_time_vacancies'
      part_time? ? update(vacancy_status: :vacancies) : update(vacancy_status: :no_vacancies)
    else
      raise InvalidVacancyStatusDescriptionError, description
    end

```

On the subject of those methods:

- Both `!` and `?` method variants are a [pretty well-documented](https://api.rubyonrails.org/classes/ActiveRecord/Enum.html) part of Rails' enum feature.
- Their conciseness can be useful in complex statements like the above.

However, happy to get a second opinion on which form is clearer/easier to understand. Let me know!

Reply:
I'm also less familiar with the bang enums but now that I know they exist I like them, the code is shouty but in a friendly kind of way ("yay vacancies!")

Reply:
@malcolmbaig first one looks good! 👍 

(A version of option 2 could use a method that returns `:vacancies` / `:no_vacancies` and update it in a single place, but this is fine as well).

@tvararu lol, yeah it's like taking a ActiveRecord object by the collar and shouting VACANCIES! at it until it complies.